src/services/TestResultsService.ts

```diff
@@ -109,52 +125,126 @@ export class TestResultsService {
     return testResults;
   }
 
-  public insertTestResult(payload: ITestResultPayload) {
-    let validationSchema = null;
-    if (Object.keys(payload).length === 0) { // if empty object
-      return Promise.reject(new HTTPError(400, ERRORS.PayloadCannotBeEmpty));
+  public updateTestResult(systemNumber: string, payload: ITestResult, msUserDetails: IMsUserDetails) {
+    this.removeNonEditableAttributes(payload);
+    let validationSchema = this.getValidationSchema(payload.vehicleType, payload.testStatus);
+    validationSchema = validationSchema!.keys({countryOfRegistration: Joi.string().valid(countryOfRegistration).required().allow("", null)});
+    validationSchema = validationSchema!.optionalKeys(["testStationType", "testerEmailAddress", "testEndTimestamp", "systemNumber", "vin"]);
+    const validation: ValidationResult<any> | any | null = Joi.validate(payload, validationSchema);
+
+    if (validation !== null && validation.error) {
+      return Promise.reject(new HTTPError(400,
+          {
+            errors: validation.error.details.map((detail: { message: string; }) => {
+              return detail.message;
+            })
+          }));
     }
+    return this.testResultsDAO.getBySystemNumber(systemNumber)
+        .then((result) => {
+          const response: ITestResultData = {Count: result.Count, Items: result.Items};
+          const testResults = this.checkTestResults(response);
+          const oldTestResult = this.getTestResultToArchive(testResults, payload.testResultId);
+          const newTestResult: ITestResult = cloneDeep(oldTestResult);
+          newTestResult.testVersion = TEST_VERSION.CURRENT;
+          oldTestResult.testVersion = TEST_VERSION.ARCHIVED;
+          mergeWith(newTestResult, payload);
+          this.setAuditDetails(newTestResult, oldTestResult, msUserDetails);
+          newTestResult.testResultId = uuidv4();
+          return this.testResultsDAO.updateTestResult(newTestResult, oldTestResult)
+              .then((data) => {
+                return newTestResult;
+              }).catch((error) => {
+                throw new HTTPError(500, error.message);
+              });
+        }).catch((error) => {
+          throw new HTTPError(error.statusCode, error.body);
+        });
+  }
 
-    switch (payload.vehicleType + payload.testStatus) {
+  public getTestResultToArchive(testResults: ITestResult[], testResultId: string): ITestResult {
+    testResults = testResults.filter((testResult) => {
```

Comment:
no "Current" checks?

Reply:
True, need to check for current as well

Reply:
resolving this current is being checked now @KHodderDVSA 
Thanks

Reply:
resolving since the CURRENT check has been added, thanks @KHodderDVSA